### ISSUE TITLE
React adds extra requirement

### ISSUE THREAD

**Issue by soniakeys** [exercism/xgo#189]
This issue continues discussion that started at https://github.com/exercism/xgo/pull/188.

When I tried solving this exercise I was bothered by the final requirement of the the test program, that a callback on a cell only be called once when both of the cell's inputs change.  My issue is that my reading of the readme not only does not make this requirement, but even worse suggests an implementation that does not work with this requirement.  The readme says "...allow a change to one value to automatically propagate to other values..." and then ..."allow for registering callbacks which get called when the value of a cell changes."  This to me indicates an implementation where changes are propagated by callbacks.

Working the exercise, when I saw the final test fail I understood its intent but I felt I had been led down a garden path.  I felt like the readme had indicated that the problem was to be implemented a certain way.  But then after coding what the readme indicated, the test program says at the very end, "sorry you did it the way the readme suggested, now here's an extra requirement tacked on that invalidates all that, so you must now code something completely different."  I felt cheated and the problem stopped being fun at that point.  Now, this happens a lot in the real world of course.  Stakeholders give you a set of requirements, then QA demands something that expands the scope of the project and that you feel was not within the intent of the original requirements.  It struck me as devious, so I wondered if that was deliberate here, following the theme of the site.

As far as surprising behavior, if I had defined a cell with two inputs and I knew both inputs were changing,  I would be alarmed to get only a single update.  It's a difference of assumptions and the readme doesn't suggest either assumption.

My preferences in order would be

1.  Remove the test program requirement that a callback be called only once when both inputs change.
2.  Update the readme to be more explicit and less potentially misleading.
3.  Leave the problem devious.

And with this choices 2 or 3, perhaps add a new test case where changes to single inputs would produce a different result but changes to both inputs leave the result the same.  Presumably you would not want a callback in this case.  Like with `b = a+1`, `c = a-1`, `d = b-c`, `d` should always be 2 and a callback attached to `d` should not be called when `a` changes.

**Comment by petertseng** [comment]
I understand I lack the authority to decide, but I will post my input in case it can be helpful.
While it might appear that I argue very strongly for a particular solution and have many arguments for the solution I support, I understand that I have no personal stake in the matter and that ultimately I would defer to your reasoned decision.

Post is long.
Points I feel are important are bolded; those are my tl;dr, along with my following statement:
I favor approach 2 (make readme explicit), I would be reluctant to do approach 1 (allow callbacks to be called more than once), and I think approach 3 (take no action) evades responsibility to the users of Exercism, so hopefully don't take approach 3.

I will first **agree that the semantics of callbacks are not well-defined within the readme**.
I think to best inform which of the various solutions to proceed with, there should be agreement on the semantics.
We probably have some unwritten assumptions, and we should write them down so that we can better understand the differences.
I see that you have already done this.

For my part, **my assumptions about the semantics of callbacks are as follows**:
A given call to `SetValue` on an `InputCell` produces *at most one* call to every callback.
More specifically after all values changes have been propagated from the `SetValue`, for each `ComputeCell`:
If the `SetValue` call caused that `ComputeCell`'s value to change, each callback on that `ComputeCell` should be called once, passing the new value.
Otherwise, the callbacks on that `ComputeCell` should not be called.

I argue in favor of this semantics in #188 (it's an external interface so the semantics should be geared toward external clients, and if an internal implementation requires a different semantics, they should use an internal callback system, not the external one).

I have two more points to add in support, and your example is very instructive so I'll use it.
I'm going to rename the cells for a bit of clarity, as I will add to the example.
Let's say we have `input` with initial value 0, and `plus1_1 = input + 1, minus1_1 = input - 1, output1 = plus1_1 - minus1_1`.
We know by looking at this that for any given value of `input`, the value of `output1` will always be 2.
I want to point out **two things that could happen if callbacks *should* perceive intermediate values** (in other words, if callbacks may be called more than once per `SetValue` call).

1. **Callbacks may perceive "impossible" values**.

Let's say that we set up this reactor and register a callback on `output1` that simply says "output changed value to (value)".
Then we perform `input.SetValue(5)`.
Depending on the order in which the updates are propagated through the system, we may see two different things printed out:

* "output changed value to 7" followed by "output changed value to 2", if `plus1_1` updates before `minus1_1`.
* "output changed value to -3" followed by "output changed value to 2", if the other way around.

The possible 7 and -3 are phantom outputs that will never exist in this system.
There is absolutely no input value that we can provide that will cause the output to take on this value.
Yet they got printed out, with potentially confusing consequences ("What? How did these values get there? The value should be 2. Is there a bug in the system?").

2. **The values perceived by callbacks may become implementation-dependent**.

Let's say that we have the above example, but now we add a few more cells: `output2 = -minus1_1 + plus1_1, minus1_2 = input - 1, plus1_2 = input + 1, output3 = plus1_2 - minus1_2`.
So, compared to `output1`, `output2` would have the `minus1_1` as the first input to its compute function, whereas `output1`s has `plus1_1` as the first input.
`output3` has the plus and minus cells in the same order in the function call, but the minus cell was created before the plus cell, whereas for `output1` the plus cell was created first.

Again we put callbacks on all three `output` cells that print out the value, the input starts out with value 0, we perform `input.SetValue(3)`.

I see a distinct possibility that many implementations will cause the callbacks to print 7, 2 for some outputs and -3, 2 on the others, despite the three output cells really having computed the same thing.
So the values perceived by callbacks are going to be dependent on either 1) which cell is the first argument to the Compute2 cell's compute function or 2) which cell got created first.

This means that tests on callbacks would have to be able to deal with this non-determinism.
(That is not difficult to do because we can enumerate the possibilities, but it needs to be pointed out)

To me it looks like both these consequences mean the reactor would be a very leaky abstraction.

Regardless of what is finally decided for semantics, **I believe the readme should contain the semantics. Any behavior that the tests check for must also be specified in the readme**.
If someone told me to implement the system described in the readme which does not specify the callback semantics, I believe I would be within my rights to ask for the semantics before proceeding.

I will sympathize on being led to barking up the wrong tree at first.
It is frustrating to discover a gotcha that requires a complete rewrite.
I will admit that I ran into the same when implementing my solution to this exercise:
I attempted to do it with channels and non-blocking propagation before realizing that the tests were not really compatible with this approach.
Reluctantly, I scrapped my approach and moved to a synchronous model wth blocking propagation.

I would like to take this opportunity to ask a sort of philosophical question about the site and why it exists.

> It struck me as devious, so I wondered if that was deliberate here, following the theme of the site.

I realize that the tagline of Exercism is "the devil is in the details".
Despite this, I don't think any problems should be *devilish* to implement.
By that, I believe I mean **no exercise should be frustrating**.
*Challenging* perhaps (maybe requiring advance planning of a design before writing code), but not *frustrating* (perhaps a hidden trap that springs and invalidates a well-planned design).
If the requirements were specified in the readme and I simply failed to understand them, well that's on me and I should endeavor to read the readme better next time.
If the requirements were not specified in the readme, I would contend that that is a bug in the readme.
It is unreasonable to hold people accountable for what they have not been told.
I acknowledge that I am relatively new to this site, and I cannot say for sure what philosophy should be followed for user experience when solving problems.
But this is what seems reasonable to me.

**Comment by soniakeys** [comment]
Peter, thank you for that detailed response.  PTAL at my PR on x-common.  It is far from completely specifying the semantics of the current Go test program but I think I added enough detail that it would have prevented my frustration in working the exercise.

**Comment by kytrinyx** [comment]
That has been merged. Are we good to close this issue?

**Comment by petertseng** [comment]
Looks like we want one more thing first:

> add a new test case where changes to single inputs would produce a different result but changes to both inputs leave the result the same

I am glad to do this but won't belable to begin work on it for about 12 hours. if anyone wish to do it before me just say the word

**Comment by kytrinyx** [comment]
Aaah, good catch. Yepp, I'll leave it open, then.

**Comment by soniakeys** [comment]
I think that new test would be nice.  Also I have suggestions for a couple more changes.  These aren't directly related to this issue but might be included in the same PR anyway.  First, my posted iteration 2 has a bug in it where it messes up the callback handles on delete.  If you add two callbacks to the same cell, then delete them in the same order it will crash.  A regression test for that would be nice.  This delta for example is enough to show my bug:

```

--- a/react/react_test.go
+++ b/react/react_test.go
@@ -172,7 +172,7 @@ func TestCallbackAddRemove(t *testing.T) {
                observed1 = append(observed1, v)
        })
        var observed2 []int
-       c.AddCallback(func(v int) {
+       cb2 := c.AddCallback(func(v int) {
                observed2 = append(observed2, v)
        })
        i.SetValue(2)
@@ -190,6 +190,7 @@ func TestCallbackAddRemove(t *testing.T) {
        if len(observed2) != 2 || observed2[1] != 4 {
                t.Fatalf("observed2 not properly called after first callback removal")
        }
+       c.RemoveCallback(cb2)
 }
```

Then second, I know I'm the one that merged it, but `assertCellValue` of commit https://github.com/petertseng/xgo/commit/17c1055a0e5685e6f8e23014bd2ca9e022255775 degrades error reporting somewhat.  A common function like this loses the line number in the test program where the error was detected.  The error message generated by Testing.Fatalf now has a line number in this common function which no longer identifies the location of the test case that failed.  One solution is to just make the `explanation` text identify the name of the failing test function or test case or whatever is needed to tell the solver where to look in the test program.  Then, I know I've dealt with this before and I think I implemented something fancy on a problem or two.  I don't remember which though or what I did.  Exercises with common test data have a similar problem.  The solution there sometimes incorporates a unique test case label or description but more generally outputs the name of the test or the name of the failing function and all input values in addition to expected and actual.

**Comment by petertseng** [comment]
> If you add two callbacks to the same cell, then delete them in the same order it will crash.

Understood, I can certainly do that.

> A common function like this loses the line number in the test program where the error was detected.

The irony that in my attempts to make error reporting better in one way (reporting expected value) I make them worse in a different way (in reporting line number)! Guilty as charged of course, I see that the tests all report line 30 as failure. The one saving grace is that it does show the name of the test function:

```
--- FAIL: TestSetInput (0.00s)
	react_test.go:30: i.Value() doesn't match initial value: expected 1, got 2
--- FAIL: TestBasicCompute1 (0.00s)
	react_test.go:30: c.Value() isn't properly computed based on initial input cell value: expected 2, got 4
... (other output omitted)
```

But we definitely agree it makes sense to make the line numbers good. I will look for a ways to address this. If worse comes to worse I will have to revert that common function altogether, but I hope I can find a solution. If you find a way before I do, then let me know.

**Comment by soniakeys** [comment]
Oh nice, that would be very cool.  Oh wait...grep...there.  robot-simulator.  I told you I tried something fancy once.  See what's going on in robot-simulator.  (If it looks bad, let me know and I'll take a look myself.)

**Comment by soniakeys** [comment]
Resolved with PRs https://github.com/exercism/x-common/pull/122 and https://github.com/exercism/xgo/pull/192.  Thank you @petertseng :)
